### PR TITLE
Add Sequence Support For MX Tables

### LINKED ISSUE #794: Add worker sequence number generation logic
To generate unique sequence numbers for bigserial columns from workers, generate sequence numbers as: [16-bit group ID][48-bit sequence number] by setting the minvalue and maxvalue on each sequence on the worker. Monotonic worker-generated sequence numbers are not and probably never will be supported. Tuples can be ordered by a (timestamp,sequence number) tuple if a consistent chronologic order is required.

### PR DESCRIPTION
**Pull request by aamederen**
Fixes #794 

This change adds support for serial columns to be used with MX tables.
Prior to this change, sequences of serial columns were created in all
workers (for being able to create shards) but never used. With MX, we
need to set the sequences so that sequences in each worker create
unique values. This is done by setting the MINVALUE, MAXVALUE and
START values of the sequence.

Steps:

- [x] Migrate the logic for setting MIN/MAX values of a sequence on a worker
- [x] Modify the existing sequences (for `start_metadata_sync_to_node` to a node which already contains the sequence for shards)
- [x] Ensure that correct owner is set for the sequences on workers
- [x] Ensure that after `stop_metadata_sync_to_node` is called,  `DROP TABLE` still works

### CODE REVIEW

File: src/backend/distributed/citus--6.1-12--6.1-13.sql [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+SET search_path = 'pg_catalog';
+
+CREATE FUNCTION master_dist_local_group_cache_invalidate()
+    RETURNS trigger
+    LANGUAGE C
+    AS 'MODULE_PATHNAME', $$master_dist_local_group_cache_invalidate$$;
+COMMENT ON FUNCTION master_dist_local_group_cache_invalidate()
+    IS 'register node cache invalidation for changed rows';
+    
+CREATE TRIGGER dist_local_group_cache_invalidate
```
Comment (marcocitus) [inline src/backend/distributed/citus--6.1-12--6.1-13.sql]: Could you do these as a separate commit?

File: src/backend/distributed/master/master_delete_protocol.c [hunk 2]
```diff
@@ -268,8 +268,13 @@ master_drop_sequences(PG_FUNCTION_ARGS)
 		appendStringInfo(dropSeqCommand, " %s", TextDatumGetCString(sequenceText));
 	}
 
-	SendCommandToWorkers(ALL_WORKERS, DISABLE_DDL_PROPAGATION);
-	SendCommandToWorkers(ALL_WORKERS, dropSeqCommand->data);
+	if (dropSeqCommand->len != 0)
+	{
+		appendStringInfoString(dropSeqCommand, " CASCADE");
```
Comment (marcocitus) [inline src/backend/distributed/master/master_delete_protocol.c]: Why do we need this?

File: src/backend/distributed/worker/worker_data_fetch_protocol.c [hunk 3]
```diff
@@ -450,6 +451,52 @@ worker_apply_inter_shard_ddl_command(PG_FUNCTION_ARGS)
...
+				(errmsg("must call worker_apply_sequence_command with a CREATE"
+						" SEQUENCE command string")));
+	}
+
+	createSequenceStatement = (CreateSeqStmt *) commandNode;
+	sequenceName = createSequenceStatement->sequence->relname;
+	sequenceSchema = createSequenceStatement->sequence->schemaname;
+
+	/* run the CREATE SEQUENCE command */
+	ProcessUtility(commandNode, commandString, PROCESS_UTILITY_TOPLEVEL,
```
Comment (marcocitus) [inline src/backend/distributed/worker/worker_data_fetch_protocol.c]: Just for my understanding, we wouldn't normally create a sequence here right? Because we create shard placements (which include the sequence) before we get here.

File: src/backend/distributed/relay/relay_event_utility.c [hunk 4]
```diff
@@ -473,6 +476,80 @@ RelayEventExtendNamesForInterShardCommands(Node *parseTree, uint64 leftShardId,
...
+	startValue = (((int64) GetLocalGroupId()) << 48) + 1;
+	appendStringInfo(startNumericString, "%lu", startValue);
+	startFloatArg = (Node *) makeFloat(startNumericString->data);
+
+	maxValue = startValue + ((int64) 1 << 48);
+	appendStringInfo(maxNumericString, "%lu", maxValue);
+	maxFloatArg = (Node *) makeFloat(maxNumericString->data);
+
+	SetDefElemArg(alterSequenceStatement, "start", startFloatArg);
+	SetDefElemArg(alterSequenceStatement, "restart", startFloatArg);
```
Comment (marcocitus) [inline src/backend/distributed/relay/relay_event_utility.c]: Do we restart the sequence when we run `start_metadata_sync_to_node`?

File: src/include/distributed/metadata_sync.h [hunk 5]
```diff
@@ -30,12 +30,15 @@ extern char * NodeListInsertCommand(List *workerNodeList);
...
 extern char * ColocationIdUpdateCommand(Oid relationId, uint32 colocationId);
+extern char * CreateSchemaDDLCommand(Oid schemaId);
 
 
 #define DELETE_ALL_NODES "TRUNCATE pg_dist_node"
 #define REMOVE_ALL_CLUSTERED_TABLES_COMMAND \
 	"SELECT worker_drop_distributed_table(logicalrelid) FROM pg_dist_partition"
 #define DISABLE_DDL_PROPAGATION "SET citus.enable_ddl_propagation TO 'off'"
+#define WORKER_APPLY_SEQUENCE_COMMAND "SELECT worker_apply_sequence_command ('%s')"
+#define WORKER_GRANT_SEQUENCE_PERMISSION_COMMAND "GRANT "
```
Comment (marcocitus) [inline src/include/distributed/metadata_sync.h]: This doesn't appear to be used.

File: src/include/distributed/metadata_sync.h [hunk 6]
```diff
@@ -30,12 +30,15 @@ extern char * NodeListInsertCommand(List *workerNodeList);
...
 extern char * NodeDeleteCommand(uint32 nodeId);
 extern char * ColocationIdUpdateCommand(Oid relationId, uint32 colocationId);
+extern char * CreateSchemaDDLCommand(Oid schemaId);
 
 
 #define DELETE_ALL_NODES "TRUNCATE pg_dist_node"
 #define REMOVE_ALL_CLUSTERED_TABLES_COMMAND \
 	"SELECT worker_drop_distributed_table(logicalrelid) FROM pg_dist_partition"
 #define DISABLE_DDL_PROPAGATION "SET citus.enable_ddl_propagation TO 'off'"
+#define WORKER_APPLY_SEQUENCE_COMMAND "SELECT worker_apply_sequence_command ('%s')"
```
Comment (marcocitus) [inline src/include/distributed/metadata_sync.h]: We shouldn't have quotes in here, and instead use parameters or `quote_` functions.

File: src/backend/distributed/metadata/metadata_sync.c [hunk 7]
```diff
@@ -689,6 +707,151 @@ MarkNodeHasMetadata(char *nodeName, int32 nodePort, bool hasMetadata)
...
+	}
+
+	shouldSyncMetadata = ShouldSyncTableMetadata(tableId);
+
+	columnType = TypeOfColumn(tableId, (int16) columnId);
+
+	if (columnType != INT8OID && shouldSyncMetadata)
+	{
+		ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+						errmsg("cannot create clustered table with columns "
```
Comment (marcocitus) [inline src/backend/distributed/metadata/metadata_sync.c]: probably shouldn't use "clustered table"

File: src/backend/distributed/metadata/metadata_sync.c [hunk 8]
```diff
@@ -689,6 +707,151 @@ MarkNodeHasMetadata(char *nodeName, int32 nodePort, bool hasMetadata)
...
+	{
+		aclcheck_error(ACLCHECK_NOT_OWNER, ACL_KIND_SEQUENCE,
+					   get_rel_name(sequenceOid));
+	}
+
+	shouldSyncMetadata = ShouldSyncTableMetadata(tableId);
+
+	columnType = TypeOfColumn(tableId, (int16) columnId);
+
+	if (columnType != INT8OID && shouldSyncMetadata)
```
Comment (marcocitus) [inline src/backend/distributed/metadata/metadata_sync.c]: One thing that worries me a bit here is indadvertedly disallowing serial type for non-MX tables. I'm not sure what the best way to solve this is, but I'd rather not disallow it.

File: src/backend/distributed/citus--6.1-12--6.1-13.sql (hunk 1 above)
Comment (aamederen) [inline src/backend/distributed/citus--6.1-12--6.1-13.sql]: Makes sense, extracted this part to a separate commit.

File: src/backend/distributed/master/master_delete_protocol.c (hunk 2 above)
Comment (aamederen) [inline src/backend/distributed/master/master_delete_protocol.c]: Basically, we send the `DROP SEQUENCE` command to all workers but `DROP TABLE` command to metadata workers. If a `stop_metadata_sync_to_node` is followed by a `DROP TABLE`, then the distributed table won't be dropped and when it comes to `DROP SEQUENCE`, the command will error out becuase there exists a table depending on the sequence.

Comment (aamederen) [inline src/backend/distributed/master/master_delete_protocol.c]: There are multiple solutions to avoid this problem:
- Try to clean a worker on `stop_metadata_sync_to_node`
- Propagate `DROP TABLE` to all workers, instead of just metadata workers
- `DROP SEQUENCE` with `CASCADE` option

File: src/backend/distributed/worker/worker_data_fetch_protocol.c (hunk 3 above)
Comment (aamederen) [inline src/backend/distributed/worker/worker_data_fetch_protocol.c]: Correct. However, it is not always the case because you may want to propagate metadata to just added worker which does not have any shards. I believe we can improve this logic within the context of #1097

File: src/backend/distributed/relay/relay_event_utility.c (hunk 4 above)
Comment (aamederen) [inline src/backend/distributed/relay/relay_event_utility.c]: Yes, the function first creates the sequence if not exists, then restarts the sequence with its proper value.

File: src/include/distributed/metadata_sync.h (hunk 6 above)
Comment (aamederen) [inline src/include/distributed/metadata_sync.h]: You are right, fixed that.

File: src/include/distributed/metadata_sync.h (hunk 5 above)
Comment (aamederen) [inline src/include/distributed/metadata_sync.h]: 😅 , removed that :)

File: src/backend/distributed/metadata/metadata_sync.c (hunk 8 above)
Comment (aamederen) [inline src/backend/distributed/metadata/metadata_sync.c]: You are right. We can have multiple ways of solving this:

- Handle serial columns with a similar approach: 24 bit space for each worker and 8 bit for worker ids. We may give a warning about the possible problems and also check about nodeid overflows.
- Postpone erroring out to metadata propagation (when we actually have to run `worker_apply_sequence_command`)
- Don't mark tables with SERIAL columns as MX tables
- Document that people should user BIGSERIAL instead of regular SERIAL, if additional 4 bytes is not that critical.

File: src/backend/distributed/metadata/metadata_sync.c (hunk 7 above)
Comment (aamederen) [inline src/backend/distributed/metadata/metadata_sync.c]: Corrected the phrase

**Comment by aamederen** [comment]
Hey @marcocitus, I have updated the code according to your comments. Could you please explain the problem you mentioned in your comment (https://github.com/citusdata/citus/issues/794#issuecomment-263812680) about the ownership problem?

**Comment by marcocitus** [comment]
If we create a sequence during `start_metadata_sync_to_node` the owner is superuser, but that means that even the table owner might not be able to use the sequence.

File: src/backend/distributed/relay/relay_event_utility.c (hunk 4 above)
Comment (marcocitus) [inline src/backend/distributed/relay/relay_event_utility.c]: So far, we've assumed it's always safe to rerun `start_metadata_sync_to_node`, though in this case it would not be. What if we remove the restart?

File: src/backend/distributed/master/master_delete_protocol.c (hunk 2 above)
Comment (marcocitus) [inline src/backend/distributed/master/master_delete_protocol.c]: ok, cascading seems fair then.

File: src/backend/distributed/relay/relay_event_utility.c (hunk 4 above)
Comment (aamederen) [inline src/backend/distributed/relay/relay_event_utility.c]: You are correct, however it requires the next value to be between minvalue and maxvalue, which is not the case unless you specify a restart. Maybe I can try checking the current value and apply the restart if the current value is greater than the min value.

Comment (aamederen) [inline src/backend/distributed/relay/relay_event_utility.c]: I have updated the part so that it restarts only if necessary. Note that since we call nextval to get the most recent value of the sequence, each `start_metadata_sync_to_node` will increment the sequence by one, but I think it is not a deal breaker.

File: src/backend/distributed/metadata/metadata_sync.c (hunk 8 above)
Comment (marcocitus) [inline src/backend/distributed/metadata/metadata_sync.c]: > Postpone erroring out to metadata propagation (when we actually have to run worker_apply_sequence_command)

I quite like this option, if we also check whether there are MX workers.

Another alternative I'd be fine with is to not adjust the ranges for serial and smallserial (send a vanilla create sequence command instead of worker_apply_sequence_command).

**Comment by aamederen** [comment]
> If we create a sequence during start_metadata_sync_to_node the owner is superuser, but that means that even the table owner might not be able to use the sequence.

I see the problem. We send a `GRANT` statement after `worker_apply_sequence_command` now.

File: src/backend/distributed/metadata/metadata_sync.c (hunk 8 above)
Comment (aamederen) [inline src/backend/distributed/metadata/metadata_sync.c]: >Postpone erroring out to metadata propagation (when we actually have to run worker_apply_sequence_command)
I quite like this option, if we also check whether there are MX workers

Done this way

File: src/backend/distributed/metadata/metadata_sync.c [hunk 9]
```diff
@@ -689,6 +708,163 @@ MarkNodeHasMetadata(char *nodeName, int32 nodePort, bool hasMetadata)
 
 
 /*
+ * SequenceDDLCommandsForTable returns a list of commands which create sequences (and
+ * their schemas) to run on workers before creating the relation. The sequence creation
+ * commands are wrapped with a `worker_apply_sequence_command` call, which sets the
+ * sequence space uniquely for each worker.
+ */
+List *
+SequenceDDLCommandsForTable(Oid relationId)
```
Comment (marcocitus) [inline src/backend/distributed/metadata/metadata_sync.c]: We only call this during metadata propagation, right? Might be worth adding a comment.

File: src/backend/distributed/metadata/metadata_sync.c [hunk 10]
```diff
@@ -689,6 +708,163 @@ MarkNodeHasMetadata(char *nodeName, int32 nodePort, bool hasMetadata)
...
+		Oid sequenceOid = (Oid) lfirst_oid(listCell);
+		char *sequenceDef = pg_get_sequencedef_string(sequenceOid);
+		char *escapedSequenceDef = quote_literal_cstr(sequenceDef);
+		StringInfo wrappedSequenceDef = makeStringInfo();
+		StringInfo sequenceGrantStmt = makeStringInfo();
+		Oid schemaId = InvalidOid;
+		char *createSchemaCommand = NULL;
+		char *sequenceName = get_rel_name(sequenceOid);
+
+		if (hasMetadataWorkers)
```
Comment (marcocitus) [inline src/backend/distributed/metadata/metadata_sync.c]: This function is a bit weird since it doesn't have any result if there are metadata workers. Is that intentional? Would it make sense to only call this function if there are metadata workers?

File: src/backend/distributed/metadata/metadata_sync.c [hunk 11]
```diff
@@ -689,6 +708,163 @@ MarkNodeHasMetadata(char *nodeName, int32 nodePort, bool hasMetadata)
...
+	Oid tableId = InvalidOid;
+	Oid columnType = InvalidOid;
+	int32 columnId = 0;
+	bool shouldSyncMetadata = false;
+	bool sequenceOwned = false;
+
+	sequenceOwned = sequenceIsOwned(sequenceOid, &tableId, &columnId);
+	if (!sequenceOwned)
+	{
+		aclcheck_error(ACLCHECK_NOT_OWNER, ACL_KIND_SEQUENCE,
```
Comment (marcocitus) [inline src/backend/distributed/metadata/metadata_sync.c]: this aclcheck is slightly weird since sequenceIsOwned is about whether the sequence is "owned" by / part of a table, rather than owned by a user.

File: src/backend/distributed/metadata/metadata_sync.c [hunk 12]
```diff
@@ -689,6 +708,163 @@ MarkNodeHasMetadata(char *nodeName, int32 nodePort, bool hasMetadata)
...
+
+	Relation pgAttribute = heap_open(AttributeRelationId, AccessShareLock);
+
+	ScanKeyInit(&scanKey[0],
+				Anum_pg_attribute_attrelid,
+				BTEqualStrategyNumber, F_OIDEQ,
+				ObjectIdGetDatum(tableId));
+	ScanKeyInit(&scanKey[1],
+				Anum_pg_attribute_attnum,
+				BTEqualStrategyNumber, F_OIDEQ,
```
Comment (marcocitus) [inline src/backend/distributed/metadata/metadata_sync.c]: attnum is int16, so shouldn't be F_OIDEQ

File: src/backend/distributed/metadata/metadata_sync.c [hunk 13]
```diff
@@ -689,6 +708,163 @@ MarkNodeHasMetadata(char *nodeName, int32 nodePort, bool hasMetadata)
...
+	const int scanKeyCount = 2;
+	ScanKeyData scanKey[scanKeyCount];
+	SysScanDesc scanDescriptor = NULL;
+	Oid returnType = InvalidOid;
+	HeapTuple tup = NULL;
+	Form_pg_attribute attrForm = NULL;
+
+	Relation pgAttribute = heap_open(AttributeRelationId, AccessShareLock);
+
+	ScanKeyInit(&scanKey[0],
```
Comment (marcocitus) [inline src/backend/distributed/metadata/metadata_sync.c]: I think this function should just be something like:
```
TupleDesc tupleDescriptor = RelationGetDescr(tableId);
Form_pg_attribute attrForm = tupleDescriptor->attrs[columnId - 1];
Oid returnType = attrFrom->atttypid;
```

File: src/backend/distributed/metadata/metadata_sync.c [hunk 14]
```diff
@@ -704,3 +880,71 @@ TruncateTriggerCreateCommand(Oid relationId)
...
+		ownerName = GetUserNameFromId(ownerId, false);
+	}
+	else
+	{
+		/* if the schema is not found, then return the name of current user */
+		ownerName = GetUserNameFromId(GetUserId(), false);
+	}
+
+	systable_endscan(scanDescriptor);
+	heap_close(namespaceRelation, NoLock);
```
Comment (marcocitus) [inline src/backend/distributed/metadata/metadata_sync.c]: This function could be a lot simpler, see pg_class_ownercheck in aclckh.c

File: src/backend/distributed/worker/worker_data_fetch_protocol.c [hunk 15]
```diff
@@ -1244,3 +1294,100 @@ check_log_statement(List *statementList)
...
+
+	SetDefElemArg(alterSequenceStatement, "start", startFloatArg);
+	SetDefElemArg(alterSequenceStatement, "minvalue", startFloatArg);
+	SetDefElemArg(alterSequenceStatement, "maxvalue", maxFloatArg);
+
+	/*
+	 * Restart the sequence if the current value of the sequence is outside of the newly
+	 * set min and max values.
+	 */
+	if (sequenceNextvalue < startValue || sequenceNextvalue > maxValue)
```
Comment (marcocitus) [inline src/backend/distributed/worker/worker_data_fetch_protocol.c]: This feels a little wonky. Especially because this function unexpectedly draws from the sequence. What if we just never restart? Group ID should never change without recreating the sequence.

File: src/test/regress/sql/multi_metadata_sync.sql [hunk 16]
```diff
@@ -380,37 +380,121 @@ SELECT logicalrelid, repmodel FROM pg_dist_partition WHERE logicalrelid = 'mx_te
...
+\c - - - :worker_2_port
+CREATE USER mx_user;
+
+\c - mx_user - :master_port
+CREATE TABLE mx_table (a int, b BIGSERIAL);
+SET citus.shard_replication_factor TO 1;
+SELECT create_distributed_table('mx_table', 'a');
+
+\c - postgres - :master_port
+SELECT master_add_node('localhost', :worker_2_port);
```
Comment (marcocitus) [inline src/test/regress/sql/multi_metadata_sync.sql]: Is it important that we do master_remove_node and master_add_node given it touches a lot of tests? Alternatively, could we add a comment on what we're actually testing here? :)

File: src/backend/distributed/metadata/metadata_sync.c [hunk 17]
```diff
@@ -689,6 +708,163 @@ MarkNodeHasMetadata(char *nodeName, int32 nodePort, bool hasMetadata)
...
+			{
+				sequenceDDLList = lappend(sequenceDDLList, createSchemaCommand);
+			}
+
+			appendStringInfo(wrappedSequenceDef,
+							 WORKER_APPLY_SEQUENCE_COMMAND,
+							 escapedSequenceDef);
+
+			appendStringInfo(sequenceGrantStmt,
+							 "GRANT ALL ON SEQUENCE %s TO %s", sequenceName, ownerName);
```
Comment (aamederen) [inline src/backend/distributed/metadata/metadata_sync.c]: Suprised to see that `ALTER SEQUENCE .. OWNED BY` and `ALTER SEQUENCE .. OWNER TO` are different things 😄

File: src/test/regress/expected/multi_metadata_sync.out [hunk 18]
```diff
@@ -861,16 +861,39 @@ SELECT logicalrelid, repmodel FROM pg_dist_partition WHERE logicalrelid = 'mx_te
...
+SELECT start_metadata_sync_to_node('localhost', :worker_1_port);
+ERROR:  cannot create an mx table with columns which use sequences, but are not BIGINT
+DROP TABLE mx_table_with_small_sequence;
 SELECT start_metadata_sync_to_node('localhost', :worker_1_port);
  start_metadata_sync_to_node 
 -----------------------------
  
 (1 row)
 
+-- Create an MX table with (BIGSERIAL) sequences
```
Comment (aamederen) [inline src/test/regress/expected/multi_metadata_sync.out]: Added 👍

File: src/backend/distributed/metadata/metadata_sync.c (hunk 9 above)
Comment (aamederen) [inline src/backend/distributed/metadata/metadata_sync.c]: Added a mention about where it is used

File: src/backend/distributed/metadata/metadata_sync.c (hunk 10 above)
Comment (aamederen) [inline src/backend/distributed/metadata/metadata_sync.c]: The place of this check was a bit weird. I moved the check at the beginning, since if there are no metadata workers, no need to run the rest of the code. However, I still like this check here because it would require double check (for calls from `start_metadata_sync_to_node` and `create_distributed_table`).

File: src/backend/distributed/metadata/metadata_sync.c (hunk 11 above)
Comment (aamederen) [inline src/backend/distributed/metadata/metadata_sync.c]: makes sense, added `pg_class_ownercheck`. We still need to keep `sequenceIsOwned` call though, in order to fetch the table and column id

File: src/backend/distributed/metadata/metadata_sync.c (hunk 13 above)
Comment (aamederen) [inline src/backend/distributed/metadata/metadata_sync.c]: This is great. With this, I have also changed the name of the function since it can give the owner of any class.

File: src/backend/distributed/metadata/metadata_sync.c (hunk 12 above)
Comment (aamederen) [inline src/backend/distributed/metadata/metadata_sync.c]: I believe this is irrelevant now, since we removed that part :)

File: src/backend/distributed/metadata/metadata_sync.c (hunk 14 above)
Comment (aamederen) [inline src/backend/distributed/metadata/metadata_sync.c]: Changed accordingly. With this, I have also changed the name of the function since it can give the owner of any class.

File: src/backend/distributed/worker/worker_data_fetch_protocol.c (hunk 15 above)
Comment (aamederen) [inline src/backend/distributed/worker/worker_data_fetch_protocol.c]: We cannot change the min/max value of an already created sequence without a restart.

File: src/test/regress/sql/multi_metadata_sync.sql (hunk 16 above)
Comment (aamederen) [inline src/test/regress/sql/multi_metadata_sync.sql]: Added comments there :) Basically we are testing that, if metadata is synced to a node which does not have the shards (therefore the sequence), the sequence creation (with a different user) would work as expected. For that, we need to remove the node from the list. 

We could do this with a `DELETE FROM pg_dist_node` command but the tests that rely on nodeid sequence should have set it explicitly anyway. Therefore, I have changed these :)

File: src/backend/distributed/metadata/metadata_sync.c [hunk 19]
```diff
@@ -689,6 +709,142 @@ MarkNodeHasMetadata(char *nodeName, int32 nodePort, bool hasMetadata)
...
+{
+	List *sequenceDDLList = NIL;
+	List *ownedSequences = getOwnedSequences(relationId);
+	ListCell *listCell;
+	char *ownerName = TableOwner(relationId);
+	bool hasMetadataWorkers = HasMetadataWorkers();
+
+	if (!hasMetadataWorkers)
+	{
+		return sequenceDDLList;
```
Comment (marcocitus) [inline src/backend/distributed/metadata/metadata_sync.c]: Taking another look at this, I think the main point of this is that we only do EnsureSupportedSequenceColumnType if hasMetadataWorkers is true, so we should probably put the check over there.

File: src/backend/distributed/metadata/metadata_sync.c [hunk 20]
```diff
@@ -689,6 +709,142 @@ MarkNodeHasMetadata(char *nodeName, int32 nodePort, bool hasMetadata)
...
+{
+	Oid tableId = InvalidOid;
+	Oid columnType = InvalidOid;
+	int32 columnId = 0;
+	bool shouldSyncMetadata = false;
+	bool sequenceOwnedByUser = pg_class_ownercheck(sequenceOid, GetUserId());
+
+	if (!sequenceOwnedByUser)
+	{
+		aclcheck_error(ACLCHECK_NOT_OWNER, ACL_KIND_CLASS, get_rel_name(sequenceOid));
```
Comment (marcocitus) [inline src/backend/distributed/metadata/metadata_sync.c]: Wait, why are we checking this here? I would expect that we already checked whether the user is the table owner at some point.

File: src/backend/distributed/worker/worker_data_fetch_protocol.c (hunk 15 above)
Comment (marcocitus) [inline src/backend/distributed/worker/worker_data_fetch_protocol.c]: Good point :). What if we just set the start? (i.e. offer no guarantees when you overflow)

File: src/backend/distributed/metadata/metadata_sync.c (hunk 20 above)
Comment (aamederen) [inline src/backend/distributed/metadata/metadata_sync.c]: I believe so. Is it possible to make the user owner of the table, but not the sequence?

File: src/backend/distributed/worker/worker_data_fetch_protocol.c (hunk 15 above)
Comment (aamederen) [inline src/backend/distributed/worker/worker_data_fetch_protocol.c]: Setting the START value has no effect without a restart :)

File: src/backend/distributed/metadata/metadata_sync.c (hunk 19 above)
Comment (aamederen) [inline src/backend/distributed/metadata/metadata_sync.c]: Moved the `hasMetadataWorkers` check inside `EnsureSupportedSequenceColumnType` function

File: src/backend/distributed/worker/worker_data_fetch_protocol.c (hunk 15 above)
Comment (marcocitus) [inline src/backend/distributed/worker/worker_data_fetch_protocol.c]: Could we do this in a different way? (E.g. check if start is changed by this command)

Comment (aamederen) [inline src/backend/distributed/worker/worker_data_fetch_protocol.c]: I have modified in a way that it checks the previously set values and alters the sequence if necessary.

**Comment by metdos** [comment]
@marcocitus, I think this is on your plate again (just a note for the greater audience).
